test/draw_test.py

```diff
@@ -1,7 +1,204 @@
+
+            draw.ellipse(surface, color, (left, top, width - 1, height - 1),
+                         border_width)
+
+            borders = get_border_values(surface, width, height)
+
+            # Check if two sides of the ellipse are touching the border
+            sides_touching = [
+                color in border for border in borders].count(True)
+            self.assertEqual(sides_touching, 2)
+
+        for width, height in sizes:
+            for border_width in (0, 1):
+                same_size(width, height, border_width)
+                for left, top in left_top:
+                    not_same_size(width, height, border_width, left, top)
+
+
+def lines_set_up():
+    """Returns the colors and surfaces needed in the tests for draw.line,
+    draw.aaline, draw.lines and draw.aalines.
+    """
+    colors = [
+            (0, 0, 0), (255, 0, 0), (0, 255, 0), (0, 0, 255), (255, 255, 0),
+            (255, 0, 255), (0, 255, 255), (255, 255, 255)]
+
+    sizes = [(49, 49), (50, 50)]
+    depths = [16, 32]
+    surfaces = []
+    for size in sizes:
+        for depth in depths:
+            # Create each possible surface type
+            surface_default = pygame.display.set_mode(size, 0, depth)
+            surface_default_SRCALPHA = pygame.Surface(size, SRCALPHA, depth)
+            surface_alpha = surface_default.convert_alpha()
+            surface_alpha_SRCALPHA = surface_default_SRCALPHA.convert_alpha()
+
+            surfaces.extend([surface_default, surface_alpha,
+                             surface_alpha_SRCALPHA, surface_alpha_SRCALPHA])
+
+    return colors, surfaces
+
+
+class DrawLineTest(unittest.TestCase):
+    """Class for testing line(), aaline(), lines() and aalines().
+    """
+
+    def test_line_color(self):
+        """|tags: ignore|
+
+        Checks if the line drawn with line_is_color() is the correct color.
+        """
+
+        def line_is_color(surface, color, draw_line):
+            """
+            Returns True if draw_line is drawn with the correct color on the
+            given surface.
+            """
+            draw_line(surface, color, (0, 0), (1, 0))
+            return surface.get_at((0, 0)) == color
+
+        for draw_line in [draw.line, draw.aaline]:
+            colors, surfaces = lines_set_up()
+            for surface in surfaces:
+                for color in colors:
+                    self.assertTrue(line_is_color(surface, color, draw_line))
+
+    def test_line_gaps(self):
+        """|tags: ignore|
+
+        Tests if the line drawn with line_has_gaps() contains any gaps.
+
+        See: #512
+        """
+
+        def line_has_gaps(surface, draw_line):
+            """Returns True if the line drawn on the surface contains gaps.
+            """
+            width = surface.get_width()
+            color = (255, 255, 255)
+
+            draw_line(surface, color, (0, 0), (width - 1, 0))
+
+            colors = [surface.get_at((width, 0)) for width in range(width)]
```

Comment:
Hello, thank you for modifications ! Some more remarks : 

1. it is not so nice to give same name for loop variable `width` and the outer `width` ; better use a single letter for loop variable in list-comprehension : 
```python
colors = [surface.get_at((x, 0)) for x in range(width)]
```
2. In the cases below, you can write `in_border.count(True) == 4` as `all(in_border)` which seems clearer to me.

3. Also, I still don't understand : for me reading the code, *line_has_gaps(* checks that there is *no* gap since _all_ are the same color ... (same remark for below *lines_have_gaps* ) ; could you please explain what the "gaps" are , or am I missing something ?

Reply:
1. Will update that.
2. Will update that.
3. Since it wasn't clear before I already created #512 showing what I meant.